### PR TITLE
deps: upgrade of SDK to 0.46 and tendermint to 0.35

### PR DESCRIPTION
**Pull request by crodriguezvega**
## Description

Upgrade of SDK to 0.46-alpha3 and Tendermint to 0.35.

This PR is based on the excellent work done by @faddat on [his PR](https://github.com/cosmos/ibc-go/pull/949). This PR is basically a subset of all the changes included in #949 (i.e. it adds only the changes strictly necessary for the upgrade of SDK and Tendermint) to narrow the scope and make the review easier.

TODO:
- Add changelog entry.
- Add migration docs.

closes: #909 #539

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Comment by fedekunze** [comment]
Will there not be a v0.45 compatible version?

**Comment by crodriguezvega** [comment]
> Will there not be a v0.45 compatible version?

You mean SDK 0.45 with Tendermint 0.35? We could look into into...

File: testing/sdk_test.go
Comment (crodriguezvega) [inline testing/sdk_test.go:218]: I am a bit unsure about removing all this part...

File: go.mod
Comment (crodriguezvega) [inline go.mod:7]: I had to add this [to make the branch build on my Mac running Big Sur](https://github.com/cosmos/cosmos-sdk/issues/11364). I will take it off before merging.

File: testing/utils.go
Comment (crodriguezvega) [inline testing/utils.go:34]: I welcome suggestions for better `godoc`.

Comment (damiannolan) [reply testing/utils.go:34]: What about something like:

```
MakeCommit iterates through the provided validator set, creating a Precommit vote for each 
participant at the provided height and round. Each vote is signed and added to the VoteSet. 
Finally, the VoteSet is committed finalizing the block
```

File: modules/core/02-client/proposal_handler.go
Comment (colin-axner) [inline modules/core/02-client/proposal_handler.go:6]: why not use `v1`?

File: testing/simapp/app.go
Comment (colin-axner) [inline testing/simapp/app.go:182]:
```suggestion
	MsgServiceRouter      *authmiddleware.MsgServiceRouter
```
I'd prefer not to shorten the name. Not shortening aligns with [sdk usage](https://github.com/cosmos/cosmos-sdk/blob/d5d8738abe548f350e81eb6087440d4d13df165b/x/auth/middleware/run_msgs.go#L16)

File: modules/apps/27-interchain-accounts/controller/keeper/keeper.go
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/keeper/keeper.go:41]: Hmm, why did we give our controller keeper access to the msg service router? Seems like code we forgot to delete. Can we open an issue and include in this release?

File: modules/core/02-client/client/proposal_handler.go
Comment (colin-axner) [inline modules/core/02-client/client/proposal_handler.go:16]: Are rest handlers finally removed?

File: modules/light-clients/07-tendermint/types/update_test.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/types/update_test.go:30]:
```suggestion
	altPubKey, err := altPrivVal.GetPubKey(suite.chainA.GetContext())
```
preferably `suite.ctx` is removed in the future

or pass in "nil" if possible since the context is discarded anyways, or `context.Background()`

File: modules/light-clients/07-tendermint/types/misbehaviour_test.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/types/misbehaviour_test.go:33]: why not use `context.Background()` as done below?

File: testing/sdk_test.go
Comment (colin-axner) [reply testing/sdk_test.go:218]: did the legacy REST endpoints get removed completely? If so I think this entire file could be deleted (unsure). You should check with the SDK team (these tests are from the SDK but were moved to ibc-go because of the dependency in ibc-go). In general they shouldn't be removed unless the functionality they were testing has been removed (which requires an ACK from the SDK team)

summoning @amaurym, any idea if these tests can be removed now?

File: modules/core/middleware/middleware.go
Comment (colin-axner) [inline modules/core/middleware/middleware.go:22]:
```suggestion
func IBCTxMiddleware(IBCKeeper *keeper.Keeper) tx.Middleware {
```

File: modules/core/middleware/middleware.go
Comment (colin-axner) [inline modules/core/middleware/middleware.go:21]: This godoc should explain the functionality of the transaction middleware (dropping duplicate relays)

File: modules/core/middleware/middleware.go
Comment (colin-axner) [inline modules/core/middleware/middleware.go:1]: I'm questioning if we should call this middleware. Seems like it would be very confusing given our notion of middleware. We should just put this in the `core/keeper`

File: modules/core/middleware/middleware.go
Comment (colin-axner) [inline modules/core/middleware/middleware.go:17]: Why is this a pointer?

File: modules/core/middleware/middleware.go
Comment (colin-axner) [inline modules/core/middleware/middleware.go:119]: we can actually change our code. Our code was accounting for these three functions:
```go
if (sdkCtx.IsCheckTx() || sdkCtx.IsReCheckTx()) && !simulate {
```

DeliverTx should return nil (we should never error on deliver tx when looking for redundant txs)

**Review by colin-axner — COMMENTED** [review]
Nice start! Left several suggestions after briefly diving through a lot of the changes. I'm wondering if we should split up this update. For example, the IBC ante handler changes are technically not required to update to v0.46 (simply remove them), then we can have a followup pr which properly fixes the code. The old structure of the ante handler code doesn't make sense to just copy over given the new API

File: testing/utils.go
Comment (colin-axner) [reply testing/utils.go:34]: Did tendermint make these functions private?

File: testing/sdk_test.go
Comment (crodriguezvega) [reply testing/sdk_test.go:218]: Ok, thanks, @colin-axner. I will check.

File: testing/utils.go
Comment (crodriguezvega) [reply testing/utils.go:34]: [Looks like](https://github.com/tendermint/tendermint/blob/master/types/test_util.go#L11)...

File: testing/simapp/app.go
Comment (crodriguezvega) [reply testing/simapp/app.go:182]: You're right. I shouldn't have changed the name...

File: modules/apps/27-interchain-accounts/controller/keeper/keeper.go
Comment (crodriguezvega) [reply modules/apps/27-interchain-accounts/controller/keeper/keeper.go:41]: It is used [here](https://github.com/cosmos/ibc-go/blob/main/modules/apps/27-interchain-accounts/controller/keeper/account.go#L61).

File: modules/core/middleware/middleware.go
Comment (crodriguezvega) [reply modules/core/middleware/middleware.go:1]: It's the [name that SDK has adopted for what were previously known ante handlers](https://github.com/cosmos/cosmos-sdk/pull/10028).

**Comment by oopcode** [comment]
hey! `testing/simapp/params/amino.go` imports "github.com/cosmos/cosmos-sdk/x/auth/legacy/legacytx", which is not present in cosmos-sdk v0.46.0-alpha3.0.20220325134903-a69764f9f01b. is this intentional?

**Comment by amaurym** [comment]
> "github.com/cosmos/cosmos-sdk/x/auth/legacy/legacytx"

This has been moved to `"github.com/cosmos/cosmos-sdk/x/auth/migrations/legacytx"`

**Comment by oopcode** [comment]
> > "github.com/cosmos/cosmos-sdk/x/auth/legacy/legacytx"
> 
> This has been moved to `"github.com/cosmos/cosmos-sdk/x/auth/migrations/legacytx"`

right! just saw a reference to the old path here: https://github.com/cosmos/ibc-go/blob/34f03df4aa59935b75281cf2724a5b4ed34997be/testing/simapp/params/amino.go#L8

File: modules/core/middleware/middleware.go
Comment (colin-axner) [reply modules/core/middleware/middleware.go:1]: Yea, `core/keeper/sdk_middleware.go` sounds good to me

File: modules/core/middleware/middleware.go
Comment (colin-axner) [reply modules/core/middleware/middleware.go:17]: I have no preference, but typically keepers are not pointers. I haven't checked into why it needs to be a pointer in app.go

File: testing/sdk_test.go
Comment (faddat) [reply testing/sdk_test.go:218]: I think that they are simply gone-- at least that's my understanding.

File: modules/core/middleware/middleware.go
Comment (faddat) [reply modules/core/middleware/middleware.go:17]: Hi Colin,

I forget the reason precisely, but I do know that I broke osmosis quite badly by not making it a pointer in app.go in the short-lived osmosis v5 release.

**Comment by faddat** [comment]
hey, I'm really happy to see the progress here and also to be getting help on discerning exactly which changes were needed for v46 compatiblity. 

For our upgrade, the primary authors were @catShaark and @nguyenvuong1122000.  

In the coming week, I'll be updating the testnet code for https://github.com/notional-labs/craft to the beta1 cosmos-sdk (or possibly release/v0.46.x) and we should have some more input on the work here.

File: modules/core/02-client/proposal_handler.go
Comment (crodriguezvega) [reply modules/core/02-client/proposal_handler.go:6]: So what's the difference between `v1` and `v1beta1`? In some places we use one and in others the other...

Comment (crodriguezvega) [reply modules/core/02-client/proposal_handler.go:6]: Looks like I need `v1beta1` here because `v1` doesn't have `Handler` and `Content`...

**Comment by crodriguezvega** [comment]
> hey! `testing/simapp/params/amino.go` imports `"github.com/cosmos/cosmos-sdk/x/auth/legacy/legacytx"`, which is not present in `cosmos-sdk v0.46.0-alpha3.0.20220325134903-a69764f9f01b`. is this intentional?

Thanks for pointing this out! I just pushed a fix.

File: modules/core/02-client/proposal_handler.go
Comment (colin-axner) [reply modules/core/02-client/proposal_handler.go:6]: opened #1282

File: modules/apps/27-interchain-accounts/host/ibc_module_test.go
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/host/ibc_module_test.go:768]: hmm, the purpose of this test was to test the value returned by tendermint `ABCIResponseResultsHash`. If we mock this function, that defeats the purpose.

File: modules/core/legacy/v100/genesis_test.go
Comment (colin-axner) [inline modules/core/legacy/v100/genesis_test.go:17]:
```suggestion
	"github.com/cosmos/ibc-go/v3/modules/core/legacy/v100"
```
?

File: modules/apps/27-interchain-accounts/host/keeper/relay_test.go
Comment (crodriguezvega) [reply modules/apps/27-interchain-accounts/host/keeper/relay_test.go:10]: I am looking into reverting the uses of `v1` but I face the following problems:

- Both `GovKeeper.SetProposal` and `GovKeeper.ActivateVotingPeriod` accept now only `v1` proposal arguments.
- In `testing/simapp/app.go`, the line `paramsKeeper.Subspace(govtypes.ModuleName).WithKeyTable(govv1.ParamKeyTable())` uses also `v1` because `ParamKeyTable` is not available in `v1beta1` anymore.

Any ideas?

File: modules/apps/27-interchain-accounts/host/ibc_module_test.go
Comment (crodriguezvega) [reply modules/apps/27-interchain-accounts/host/ibc_module_test.go:768]: `ABCIResponsesResultHash` got removed in 0.35, but we are just replicating [its functionality from 0.34](https://github.com/tendermint/tendermint/blob/v0.34.19/state/store.go#L356-L358) into our own function.

File: modules/core/legacy/v100/genesis_test.go
Comment (crodriguezvega) [reply modules/core/legacy/v100/genesis_test.go:17]: Go compiler is importing it like this, I guess there's `clientv100 "github.com/cosmos/ibc-go/v3/modules/core/02-client/legacy/v100"` a bit above...

File: modules/apps/27-interchain-accounts/host/ibc_module_test.go
Comment (colin-axner) [reply modules/apps/27-interchain-accounts/host/ibc_module_test.go:768]: What was it replaced by? We will need to ensure the tx response is still included in the consensus hash, otherwise we need to remove inclusion of the transaction response in the interchain account acknowledgement

Comment (colin-axner) [reply modules/apps/27-interchain-accounts/host/ibc_module_test.go:768]: If the transaction response is not included in the consensus hash, the tests should break, but if we mock the behaviour that they are included, we will never know if tendermint changes this functionality

File: modules/apps/27-interchain-accounts/host/keeper/relay_test.go
Comment (colin-axner) [reply modules/apps/27-interchain-accounts/host/keeper/relay_test.go:10]: Got it, lets just leave it then. No need to spend engineering resources downgrading to an older version if the downgrade isn't trivial

File: modules/core/legacy/v100/genesis_test.go
Comment (colin-axner) [reply modules/core/legacy/v100/genesis_test.go:17]: Is the go complier importing like this or is it your text editor?

**Review by colin-axner — COMMENTED** [review]
Looking good! I still need to rereview the sdk middleware code and I am still concerned about updating the `ABCIResponsesResultHash`, implementing the function ourself defeats the purpose of the test added to indicate when the expected downstream behaviour changes 

I would prefer updates to newer release (ie beta3) to be done in a separate pr since this one already has many changes. Ideally we merge this as soon as the `ABCIResponsesResultHash` is resolved

**Review by colin-axner — APPROVED** [review]
utACK pending the merge of #1349 (approval only applies to merging into the 0.46 feature branch, not main)

File: go.mod
Comment (okwme) [inline go.mod:13]: could we try this with `cosmos-sdk v0.46-beta2` and see if there are any issues? that's what's being used in gaia atm

Comment (crodriguezvega) [reply go.mod:13]: yes, will do next week!

Comment (okwme) [reply go.mod]: got inspired and pushed a commit!
tests pass but not sure if it's necessary to add anything to `options.ExtensionOptionChecker` or `options.TxFeeChecker`

File: modules/apps/27-interchain-accounts/host/ibc_module_test.go
Comment (crodriguezvega) [reply modules/apps/27-interchain-accounts/host/ibc_module_test.go]: @colin-axner [refactored the tests](https://github.com/cosmos/ibc-go/pull/1349).

File: go.mod
Comment (crodriguezvega) [reply go.mod]: Thanks, @okwme! Yes, some changes will be needed there eventually when we get a `beta3` or `rc` of 0.46, but for now it should be ok.

**Comment by crodriguezvega** [comment]
Another PR will be opened when there's an SDK 0.46 RC pre-release.

**Pull request merged**
